### PR TITLE
At debuginfo=0, don't inline debuginfo when inlining

### PR DESCRIPTION
**Pull request by scottmcm**
See https://github.com/rust-lang/rust/pull/123949#issuecomment-2060493285 for info.

### CODE REVIEW

**Comment by oli-obk** [comment]
This makes a lot of sense, none of the removed debug info seemed very useful, and is a perf improvement when not having debug info enabled due to stripping debug info from libstd functions when inlining them.

r? @oli-obk 

@bors r+

File: tests/mir-opt/const_prop/inherit_overflow.main.GVN.panic-unwind.diff [hunk 1]
```diff
@@ -9,8 +9,6 @@
       scope 1 {
       }
```
Comment (klensy) [inline tests/mir-opt/const_prop/inherit_overflow.main.GVN.panic-unwind.diff:10]: Not related to PR, but this empty scopes *looks* useless, is they needed?

File: src/bootstrap/src/core/builder.rs [hunk 2]
```diff
@@ -2071,6 +2071,14 @@ impl<'a> Builder<'a> {
             // break when incremental compilation is enabled. So this overrides the "no inlining
             // during incremental builds" heuristic for the standard library.
             rustflags.arg("-Zinline-mir");
+
+            // always pass this after the next `#[cfg(bootstrap)]` update.
```
Comment (onur-ozkan) [inline src/bootstrap/src/core/builder.rs:2075]: > @onur-ozkan for bootstrap double-checking?

Thanks for the ping, just a nit suggestion from my side:

```suggestion
            // FIXME: always pass this after the next `#[cfg(bootstrap)]` update.
```

**Review by onur-ozkan — APPROVED** [review]

**Comment by oli-obk** [comment]
Hmm... why would libstd generate debug info that'll get inlined if debug info is disabled?

**Comment by oli-obk** [comment]
Thanks! Makes sense

@bors r+

**Comment by therealprof** [comment]
Whee, finally... 🕺🏻

**Comment by rylev** [comment]
Seems like [@oli-obk signed off on the perf regressions](https://github.com/rust-lang/rust/pull/123949#issuecomment-2060800859) seen here since the improvements outweighed them.

@rustbot label: +perf-regression-triaged

File: compiler/rustc_mir_transform/src/inline.rs [hunk 3]
```diff
@@ -699,7 +699,19 @@ impl<'tcx> Inliner<'tcx> {
...
         caller_body.local_decls.extend(callee_body.drain_vars_and_temps());
         caller_body.source_scopes.extend(&mut callee_body.source_scopes.drain(..));
-        caller_body.var_debug_info.append(&mut callee_body.var_debug_info);
+        if self
+            .tcx
+            .sess
+            .opts
+            .unstable_opts
+            .inline_mir_preserve_debug
+            .unwrap_or(self.tcx.sess.opts.debuginfo != DebugInfo::None)
```
Comment (saethlin) [inline compiler/rustc_mir_transform/src/inline.rs:708]: Why isn't this `== DebugInfo::Full`? Isn't this current implementation retaining variable debuginfo when compiling with `line-tables-only`?

Comment (Kobzol) [reply compiler/rustc_mir_transform/src/inline.rs:708]: Before, it was always copied during inlining, so it at least shouldn't be a regression, right? And when the debuginfo is not set to generate variable debuginfo, shouldn't `var_debug_info` be empty anyway?

Comment (saethlin) [reply compiler/rustc_mir_transform/src/inline.rs:708]: > so it at least shouldn't be a regression, right?

Yeah, this *ought not* to be a regression, but I was reading this code and this condition is surprising.

> And when the debuginfo is not set to generate variable debuginfo, shouldn't `var_debug_info` be empty anyway?

When we load `optimized_mir` for the standard library, it should always have variable debuginfo.

Comment (Kobzol) [reply compiler/rustc_mir_transform/src/inline.rs:708]: Oh, I see. So, I thought about familiarizing myself a bit with the debuginfo handling, I'll try to send a PR that improves this then.

Comment (Kobzol) [reply compiler/rustc_mir_transform/src/inline.rs:708]: https://github.com/rust-lang/rust/pull/142890 fixes this.

**Pull request merged**